### DISCUSSION TITLE
Adding multiple "program" devices

### DISCUSSION THREAD

**Discussion by ohemgipl** [artisan-roaster-scope/artisan#1278] [Q&A]
Morning~

I'm having trouble configuring multiple "program" devices. Currently, my setup functions as follows:

Primary device (ET/BT): Yocto Thermocouple
Extra Devices (HUMIDITY, SMOKE): Using the "program" function, running a python script that reads the serial output of an arduino.

I would like to add two additional sensors via arduino serial. Because the artisan help page states "you can configure an unlimited number of extra devices, each providing two additional channels", I have prepared a second arduino, that outputs readings from the two additional sensors, and an additional python script for artisan to run.

But I can't find a way to direct this second artisan device to the second python script. Artisan only ever runs the first script twice, and duplicates the results of the first device.

Is there a way to make this work? Perhaps editing the extra device config directly to point artisan to the second program?  Perhaps there is a GUI feature I'm missing somewhere?

Any help or advice would be much appreciated.

**Comment by MAKOMO** [comment]
> Morning~
> 
> I'm having trouble configuring multiple "program" devices. Currently, my setup functions as follows:
> 
> Primary device (ET/BT): Yocto Thermocouple Extra Devices (HUMIDITY, SMOKE): Using the "program" function, running a python script that reads the serial output of an arduino.
> 
> I would like to add two additional sensors via arduino serial. Because the artisan help page states "you can configure an unlimited number of extra devices, each providing two additional channels", I have prepared a second arduino, that outputs readings from the two additional sensors, and an additional python script for artisan to run.
> 
> But I can't find a way to direct this second artisan device to the second python script. Artisan only ever runs the first script twice, and duplicates the results of the first device.
> 
> Is there a way to make this work? Perhaps editing the extra device config directly to point artisan to the second program? Perhaps there is a GUI feature I'm missing somewhere?
> 
> Any help or advice would be much appreciated.

Artisan supports only one Program script for input. However, that script can deliver up to 10 readings per call. The first two are assigned to the extra device type "Program", the 3rd and 4th to the extra device type "Program 34" and so forth until the extra device type "Program 910" to fetch the 9th and 10th reading. Press the Help button next to the Program script widget in the Device Assignment dialog (menu Config >> Device) for some explanation.

**Reply by ohemgipl** [reply]
thank you for the explaination.
I will learn more about pyserial and try returning results from both ports in the same script.

**Reply by MAKOMO** [reply]
I don't understand this. Why would you use two Arduinos and two ports!? Just connect all your sensors just to one Arduino.
